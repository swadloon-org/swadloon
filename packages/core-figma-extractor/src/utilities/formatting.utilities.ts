import Case from 'case';
import prettier, { Options } from 'prettier';

const prettierConfig: Options = {
  singleQuote: true,
  tabWidth: 2,
  printWidth: 120,
};

/**
 * Format text content
 */
export function formatOutput(textContent: string, options: prettier.Options) {
  return prettier.format(textContent, { ...prettierConfig, ...options });
}

/**
 * Split numbers in camel case with '-'
 * @example info900 => info-900
 */
export function formatNameNumberKebabCase(text: string) {
  const kebabText = Case.kebab(text);
  const match = /[a-z](\d+)/gi.exec(kebabText);

  if (!match) {
    return kebabText;
  }

  const splittedText = [kebabText.slice(0, match.index + 1), kebabText.slice(match.index + 1)].join(
    '-'
  );
  return splittedText;
}
