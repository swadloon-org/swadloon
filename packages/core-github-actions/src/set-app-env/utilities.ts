import * as core from '@actions/core';

import { TEST_ENV } from '@newrade/core-common';

import { ENV } from '../../types/dot-env.js';

/**
 * Export an env variable using @actions/core's utility (env as any) for tests
 *
 * @param name name of the variable e.g. MY_VAR
 * @param value value that will be converted to a string and assigned
 */
export function exportVariable(env?: Partial<ENV>, name?: string, value?: string | null | number) {
  if (!env) {
    throw Error(`env must be passed to ${exportVariable.name}`);
  }

  if (!name) {
    throw Error(`env name must be defined got: ${name}`);
  }

  if (!value) {
    console.info(`env ${name} has value ${value}, it will be converted to empty string ''`);
  }

  if (env.TEST_ENV === TEST_ENV.CI || env.TEST_ENV === TEST_ENV.LOCAL) {
    // for test purpose we simply assign
    if (typeof value === 'undefined' || value === null) {
      (env as any)[name] = '';
      return;
    }

    (env as any)[name] = value.toString();
    return;
  }

  // not a test so we execute the assignment with the real utility
  core.exportVariable(name, `${value}`);
}

export function join(paths?: (string | undefined)[]): string {
  if (!paths) {
    return '';
  }

  if (!paths.length) {
    return '';
  }

  return paths.filter((path) => path && path.length > 1).join('.');
}
