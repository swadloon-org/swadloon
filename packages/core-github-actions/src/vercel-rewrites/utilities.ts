import { promises as fsp } from 'fs';

import * as core from '@actions/core';

import { VercelConfig } from './vercel.js';

export function readVercelConfigFile(path: string) {
  core.info(`reading vercel.json file: ${path}`);
  return fsp
    .readFile(path, 'utf-8')
    .then((result) => {
      return JSON.parse(result) as VercelConfig;
    })
    .catch((error) => {
      core.error(`could not read file ${path}`);
      return undefined;
    });
}

export function updateVercelConfigFile(path: string, data: VercelConfig) {
  core.info(`replacing vercel.json file content`);
  return fsp.writeFile(path, JSON.stringify(data, null, 2)).then(() => {
    return readVercelConfigFile(path);
  });
}
