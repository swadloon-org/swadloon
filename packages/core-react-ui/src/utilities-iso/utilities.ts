/**
 *
 * General purposes utility functions
 *
 */

/**
 * Wrapper for Objects.keys
 */
export function keys<T extends object | undefined>(object: T) {
  if (!object) {
    return [] as (keyof T)[];
  }
  return Object.keys(object) as (keyof T)[];
}

/**
 * Wrapper for Objects.values
 */
export function values<T extends object>(object: T) {
  return Object.values(object) as T[keyof T][];
}

/**
 * Convert a number into a px string.
 * @param px value in pixel
 */
export function px({ value }: { value?: number }): string {
  if (!value) {
    return '';
  }
  if (isNaN(value as number)) {
    return '';
  }
  return `${value}px`;
}

/**
 * Convert a number into a em string.
 * @param em value in pixel
 */
export function em({ value }: { value?: number }): string {
  if (!value) {
    return '';
  }
  if (isNaN(value as number)) {
    return '';
  }
  return `${value}em`;
}

/**
 * Convert a number (in px) into a rem value.
 * @param px value in pixel
 * @param baseUnitPx font-size set on the <html/> element
 */
export function pxToRem({ baseUnitPx, value }: { baseUnitPx: number; value: number }): string {
  return `${value / baseUnitPx}rem`;
}

/**
 * Convert a number (in px) into a em relative value.
 * @param px value in pixel
 * @param fontSize font-size set on the element
 */
export function pxToEm({ fontSize, value }: { fontSize: number; value: number }): string {
  return `${value / fontSize}em`;
}

/**
 * Convert px value (e.g. '15px') into a rem value.
 * @param px value in string px value
 * @param baseUnitPx font-size set on the <html/> element
 */
export function pxStringToRem({
  baseFontSize,
  value,
}: {
  baseFontSize: number;
  value: string;
}): string {
  if (!value?.length) {
    return '';
  }

  const pxPattern = /(\d+\.?(\d+))/g;
  const match = pxPattern.exec(value);
  if (!(match?.length && match[0]?.length)) {
    return '';
  }
  const conversion = Number(match[0]);
  if (!conversion) {
    return '';
  }

  return `${conversion / baseFontSize}rem`;
}

/**
 * Convert px value (e.g. '15px') into a number value.
 * @param px value in string px value
 */
export function pxStringToNumber({ value }: { value?: string }): number | undefined {
  if (!value?.length) {
    return undefined;
  }

  const pxPattern = /(\d+\.?(\d+))/g;
  const match = pxPattern.exec(value);
  if (!(match?.length && match[0]?.length)) {
    return undefined;
  }

  const conversion = Number(match[0]);
  if (isNaN(conversion)) {
    return undefined;
  }

  return conversion;
}
