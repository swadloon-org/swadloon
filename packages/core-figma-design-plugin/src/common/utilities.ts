import { pascal } from 'case';

import { VIEWPORT } from '@newrade/core-design-system';
import { keys } from '@newrade/core-react-ui/utilities-iso';

export function formatNameFigma(value?: string | (string | undefined)[]): string {
  if (!value) {
    return '';
  }

  if (Array.isArray(value)) {
    return value.map((str) => (str ? pascal(str) : '')).join('/');
  }

  return pascal(value);
}

export function isViewportProp(value?: VIEWPORT | string): boolean {
  if (!value) {
    return false;
  }

  const found = keys(VIEWPORT).find((viewport) => viewport === value);

  return !!found;
}
