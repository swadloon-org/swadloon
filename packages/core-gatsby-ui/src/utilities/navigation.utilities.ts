import { capital, kebab, lower, title } from 'case';

import { SITE_LANGUAGES } from '@newrade/core-common';
import { GatsbyMarkdownFilePageContext } from '@newrade/core-gatsb-config/config';
import { getLangSimpleCode } from '@newrade/core-react-ui/src/seo/meta.utilities';

import { NavItem } from '../navigation/nav-item.model.js';
import { Navigation } from '../navigation/navigation.model.js';

import { GatsbyPageNode } from './gatsby-page-node.js';

/**
 * Returns a Navigation object from page nodes.
 * Accepts arrays of values to sort the directories and items.
 */
export function getNavigationFromPageNodes({
  name,
  pageNodes,
  locales,
  sortOrderItems,
  sortOrderDirectories,
  excludedItems,
  uppercaseWords = ['wsl', 'ui', 'ux', 'seo', 'ssh', 'css', 'api', 'ci', 'vm', 'cms', 'pr', 'cli'],
  formatName,
  formatDisplayName,
}: {
  name: string;
  pageNodes?: GatsbyPageNode[];
  locales?: SITE_LANGUAGES[];
  sortOrderItems?: string[];
  sortOrderDirectories?: string[];
  excludedItems?: string[];
  uppercaseWords?: string[];
  formatName?: (name?: string | null) => string;
  formatDisplayName?: (name?: string | null) => string;
}): Navigation {
  if (!pageNodes?.length) {
    return {
      name,
      items: [],
    };
  }

  const normalizedSortOrderItems = sortOrderItems?.map((item) => normalizeName(item));
  const normalizedSortOrderDirectories = sortOrderDirectories?.map((item) => normalizeName(item));
  const normalizedIgnoredItems = excludedItems?.map((item) => normalizeName(item));

  const filteredPageNodes = pageNodes
    // remove 404 pages
    .filter((node) => !(/404/gi.test(node.context?.name || '') || /404/gi.test(node.path || '')))
    // keep page in specified locale
    .filter((node) => {
      if (!(locales && node.context?.locale)) {
        return node;
      }

      if (
        locales.find(
          (locale) => getLangSimpleCode(locale) === getLangSimpleCode(node.context?.locale)
        )
      ) {
        return true;
      }

      return false;
    })
    // remove ignored items
    .filter(
      (node) => !normalizedIgnoredItems?.find((item) => item === normalizeName(node.context?.name))
    );

  // find the dirNames
  const dirNames = filteredPageNodes.map(dirNamePredicate);

  function dirNamePredicate(node: GatsbyPageNode) {
    if (node.context?.name && /index/.test(node.context?.name)) {
      return getPageDirFromPathIndex(node.path);
    }

    return getPageDirFromPath(node.path);
  }

  // for item at the root the dir name is ''
  const navigation = [...new Set(dirNames)].reduce(
    (previous, current) => {
      // for each dir name, transform the nodes and place them in .items
      const currentDirName = current;

      const dirNameNodes = filteredPageNodes.filter(
        (node) => dirNamePredicate(node) === currentDirName
      );
      if (!dirNameNodes) {
        return previous;
      }

      previous.items = [
        ...previous.items,
        // create a new item
        {
          name: currentDirName ? currentDirName : 'Home',
          displayName: formatDisplayName
            ? formatDisplayName(currentDirName)
            : defaultFormatDisplayName({ name: currentDirName, uppercaseWords }),
          path: currentDirName ? currentDirName : '',
          // for each child nodes, creat a nav item
          items: dirNameNodes
            .map((node) => getNavItemShallow(node))
            // apply sorting on items, if specified
            .sort((a, b) => {
              if (!normalizedSortOrderItems) {
                return 0;
              }

              const indexA = normalizedSortOrderItems.indexOf(normalizeName(a.name));
              const indexB = normalizedSortOrderItems.indexOf(normalizeName(b.name));

              if (indexA === indexB) {
                return 0;
              }

              if (indexA === -1) {
                return 1;
              }

              if (indexB === -1) {
                return -1;
              }

              return indexA > indexB ? 1 : -1;
            }),
        },
        // apply sorting on directories, if specified
      ].sort((a, b) => {
        if (!normalizedSortOrderDirectories) {
          return 0;
        }

        const indexA = normalizedSortOrderDirectories.indexOf(normalizeName(a.name));
        const indexB = normalizedSortOrderDirectories.indexOf(normalizeName(b.name));

        if (indexA === indexB) {
          return 0;
        }

        if (indexA === -1) {
          return 1;
        }

        if (indexB === -1) {
          return -1;
        }

        return indexA > indexB ? 1 : -1;
      });

      return previous;
    },
    {
      name,
      items: [],
    } as Navigation
  );

  return navigation;

  function getNavItemShallow(node: GatsbyPageNode): NavItem {
    return {
      name: formatName ? formatName(node.context?.name) : defaultFormatName(node.context?.name),
      displayName: formatDisplayName
        ? formatDisplayName(node.context?.name || node.context?.displayName)
        : defaultFormatDisplayName({
            name: node.context?.name || node.context?.displayName,
            lang: node.context?.locale as SITE_LANGUAGES,
            uppercaseWords,
          }),
      // leave the node path untouched
      path: node.path as string,
      frontmatter: (node.context as GatsbyMarkdownFilePageContext)?.frontmatter,
    };
  }
}

/**
 * @param path relative path
 * @example /dir/page-name.tsx => dir
 */
export function getPageDirFromPath(path?: string | null): string {
  if (!path) {
    return '';
  }

  const reg =
    /(\/(?<lang>fr|en|fr_CA|en_CA|fr-CA|en-CA))?(\/?(?<dirname>.+)\/)?((?<pagename>[a-z\-]+|\/))(\.(?<ext>jsx|tsx|mdx|md))?/gi;
  const match = reg.exec(path);
  const dirName = match?.groups?.dirname;

  return dirName || '';
}

export function getPageDirFromPathIndex(path?: string | null): string {
  if (!path) {
    return '';
  }

  const reg =
    /(\/(?<lang>fr|en|fr_CA|en_CA|fr-CA|en-CA))?(\/(?<source>[a-z_-]+))(\/(?<dirname>.+)\/)/gi;
  const match = reg.exec(path);
  const dirName = match?.groups?.dirname;
  const source = match?.groups?.source;

  if (!source) {
    return '';
  }

  return dirName || '';
}

export function normalizeName(name?: string | null) {
  if (!name) {
    return '';
  }
  const nameWithoutLocale = name.replace(/^(en|fr)\./gi, '');

  return kebab(nameWithoutLocale);
}

export function defaultFormatDisplayName({
  name,
  lang,
  uppercaseWords,
}: {
  name?: string | null;
  lang?: SITE_LANGUAGES | null;
  uppercaseWords?: string[];
}): string {
  if (!name) {
    return '';
  }

  const nameWithoutLocale = name.replace(/^(en|fr)\./gi, '');
  const nameWithoutDotPage = nameWithoutLocale.replace('.page', '');
  const nameNoTrailingDash = nameWithoutDotPage.replace(/\s-\s$/, '');

  if (lang === SITE_LANGUAGES.FR || lang === SITE_LANGUAGES.FR_CA) {
    const nameDefaultIndex = nameNoTrailingDash.replace('index', 'AperÃ§u');
    const nameWithFRTitle = nameDefaultIndex
      .trim()
      .split('')
      .map((char, index) => (index === 0 ? capital(char) : char))
      .join('');
    const nameWithUppercaseWords = getNameWithUppercaseWords({
      name: nameWithFRTitle,
      words: uppercaseWords,
    });
    return nameWithUppercaseWords;
  }

  const nameDefaultIndex = nameNoTrailingDash.replace('index', 'Overview');
  const nametTitle = title(nameDefaultIndex);
  const nameWithUppercaseWords = getNameWithUppercaseWords({
    name: nametTitle,
    words: uppercaseWords,
  });
  return nameWithUppercaseWords;
}

function getNameWithUppercaseWords({ name, words }: { name?: string; words?: string[] }): string {
  if (!name) {
    return '';
  }

  if (!words?.length) {
    return name;
  }

  const wordsToFind = words.map((word) => lower(word));

  return name
    .split(' ')
    .map((part) => (wordsToFind.includes(lower(part)) ? part.toUpperCase() : part))
    .join(' ');
}

export function defaultFormatName(name?: string | null): string {
  if (!name) {
    return '';
  }
  return name
    .replace('.page', '')
    .replace('design-system-', '')
    .replace(/\s-\s$/, '');
}
