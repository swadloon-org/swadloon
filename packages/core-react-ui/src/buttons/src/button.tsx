import React, { useRef } from 'react';

import { ButtonIcon, ButtonSize, TEXT_STYLE, Variant } from '@newrade/core-design-system';
import { useCommonProps } from '@newrade/core-react-ui-hooks';

import { usePreventPinchZoom } from '../hooks/use-prevent-pinch-zoom.js';
import { IconComp } from '../icons/icon.js';
import { Primitive } from '../primitive/primitive.js';
import { Label } from '../text/label.js';
import { getDefaultTextFromProps, getMergedClassname } from '../utilities-iso.js';

import * as styles from './button.css.js';
import { ButtonAsType, ButtonCompProps } from './button.props.js';
import { getLabelSizeForButtonSize } from './button.utilities.js';

type Props = ButtonCompProps;

const defaultProps: Props = {
  as: 'button',
  variant: Variant.primary,
  size: ButtonSize.medium,
  role: 'button',
};

/**
 * Main button component
 */
export const Button = React.forwardRef<any, Props>(function Button(
  {
    role = defaultProps.role,
    children,
    as = defaultProps.as,
    AsElement,
    variant = defaultProps.variant,
    collapsePadding,
    size = defaultProps.size,
    disabled,
    loading,
    icon,
    Icon,
    iconStyle,
    IconSVG,
    ...props
  },
  forwardedRef
) {
  /**
   *
   * Ref forwarding
   *
   */

  const localRef = useRef<HTMLButtonElement>(null);
  const ref = forwardedRef ? (forwardedRef as React.RefObject<HTMLButtonElement>) : localRef;

  /**
   *
   * Event handling
   *
   */

  usePreventPinchZoom(ref.current);
  // usePreventLongPress(ref.current);

  /**
   *
   * Icon
   *
   */

  const dataicon = Icon ? (icon ? icon : ButtonIcon.right) : ButtonIcon.none;
  const iconClassNames = dataicon
    ? getMergedClassname([
        styles.iconVariants({
          position: dataicon as 'left',
          disabled: disabled,
        }),
      ])
    : '';

  //
  // If an SVG is passed, we use that, otherwise use the normal icon component
  //
  const IconSvg = IconSVG ? (
    React.cloneElement(IconSVG as React.ReactElement, {
      className: iconClassNames,
      preserveAspectRatio: `xMinYMin meet`,
    })
  ) : icon && Icon ? (
    <IconComp name={Icon} className={iconClassNames} style={iconStyle}></IconComp>
  ) : null;

  /**
   *
   * Rendering
   *
   */

  const type = as ? as : 'button';

  const collapsePaddingProp =
    dataicon === ButtonIcon.icon ? `${collapsePadding}-icon` : collapsePadding;

  const commonProps = useCommonProps<'button'>({
    role: as === 'button' ? undefined : role,
    classNames: [
      styles.variants({
        variant: variant,
        size: size,
      }),
    ],
    disabled: disabled ? true : false,
    // @ts-ignore
    dataicon: dataicon,
    datapaddingcollapse: collapsePaddingProp,
    ...props,
  });

  const renderedChildren = children
    ? children
    : dataicon === ButtonIcon.icon
    ? ' '
    : getDefaultTextFromProps('button', {
        variant,
        size,
        icon,
        disabled,
      });

  //
  // Render the label and icon
  //
  const ButtonContent = (
    <>
      {icon === ButtonIcon.left ? IconSvg : null}
      {icon === ButtonIcon.icon ? null : (
        <Label
          variantDisplay={'inline'}
          textStyle={TEXT_STYLE.bold}
          variant={getLabelSizeForButtonSize(size)}
          className={styles.labelVariants({
            alignment: 'center',
          })}
        >
          {renderedChildren}
        </Label>
      )}
      {icon === ButtonIcon.right || icon === ButtonIcon.icon ? IconSvg : null}
    </>
  );

  /**
   * Custom Element rendering
   */

  const CustomElement = AsElement
    ? React.cloneElement(
        AsElement as React.ReactElement,
        { ...commonProps, ref },
        <>{ButtonContent}</>
      )
    : null;

  if (CustomElement) {
    return CustomElement;
  }

  /**
   * Normal element rendering
   */

  return (
    <Primitive<ButtonAsType> as={type} ref={ref} {...commonProps}>
      {ButtonContent}
    </Primitive>
  );
});
