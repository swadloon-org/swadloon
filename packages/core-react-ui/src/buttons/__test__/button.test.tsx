import React from 'react';

import { renderTest } from '../../utilities-test/test-render.utilities.js';
import { Button } from '../button.js';

describe('the button component', () => {
  it('should render correctly', () => {
    const { container } = renderTest(<Button>Hello world</Button>);
    expect(container.firstChild).toMatchSnapshot();
  });
});
