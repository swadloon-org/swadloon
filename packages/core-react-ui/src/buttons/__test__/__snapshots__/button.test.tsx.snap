// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the button component should render correctly 1`] = `
<div>
  <button
    class="button__x17q3ro button_variant_primary__x17q3rp button_size_medium__x17q3rw"
    dataicon="none"
  >
    <div
      class="button__x17q3rz button_alignment_center__x17q3r10 label__1i57ob90  label_size_small__1i57ob93 label_style_bold__1i57ob96"
    >
      Hello world
    </div>
  </button>
</div>
`;
